File: #40_Combination_Sum_II/Solution.java
import java.util.*;

public class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new LinkedList();
        if (candidates == null || candidates.length == 0) return result;
        List<Integer> list = new LinkedList();
        Arrays.sort(candidates);
        backtrace(candidates, target, 0, 0, result, list);
        return result;
    }

    private void backtrace(int[] candidates, int target, int start, int sum, List<List<Integer>> result, List<Integer> list) {
    	for (int i = start; i < candidates.length; i++) {
    		/* debug */
    		// System.out.println("###");
    		// for (Integer it : list) {
    		// 	System.out.print(it + " ");
    		// }
    		// System.out.println();
    		// System.out.println("try " + candidates[i]);
    		sum += candidates[i];
    		if (sum > target) return;
    		list.add(candidates[i]);
    		if (sum == target) {
    			List<Integer> newList = new LinkedList(list);
    			// System.out.println("add!");
    			result.add(newList);
    		}
    		/* sum < target */
    		backtrace(candidates, target, i + 1, sum, result, list);
    		sum -= candidates[i];
    		/* remove(int index) */
    		list.remove(list.size() - 1);
    		while (i + 1 < candidates.length && candidates[i] == candidates[i + 1]) i++;
    	}
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {10,1,2,7,6,1,5};
    	int target = 8;
    	List<List<Integer>> result = sol.combinationSum2(t, target);
    	for (List<Integer> list : result) {
    		for (Integer it : list) {
    			System.out.print(it + " ");
    		}
    		System.out.println();
    	}
    }
}

File: #40_Combination_Sum_II/Solution.py
'''
1. If duplicates exist, need to skip the later ones to avoid duplicates in result
2. Must stop search once sum is too large or find a match
'''

class Solution:
    # @param {integer[]} candidates
    # @param {integer} target
    # @return {integer[][]}
    def combinationSum2(self, candidates, target):
        if not candidates:
        	return []
        candidates.sort() # sort the array
        re = []
        self.bt(candidates, target, [], re) # backtracing
        return re

    def bt(self, nums, target, tmp, result):
    	if not nums:
    		return
    	for x in range(len(nums)): # for each number
    		if x > 0 and nums[x] == nums[x - 1]: # skip if this one is the same as last
    			continue
    		if target < nums[x]:   # sum is too large, return because following elements are larger
    			return
    		tmp.append(nums[x])
    		 # happens to match, use list to create a hard copy and return immediately to avoid duplicates
    		 # this check must be done before entering deeper stack to avoid duplicates
    		if target == nums[x]: 
    			result.append(list(tmp))
    			tmp.pop()
    			return
    		self.bt(nums[x + 1:], target - nums[x], tmp, result) # bt
    		tmp.pop()


sol = Solution()
t= [10,1,2,7,6,1,5]
print sol.combinationSum2(t, 8)